component/attr_cache/attr_cache.go

```diff
@@ -651,61 +617,40 @@ func (ac *AttrCache) DeleteFile(options internal.DeleteFileOptions) error {
 // and search the contents of all of its ancestors,
 // to record which of them contain objects in their subtrees
 func (ac *AttrCache) updateAncestorsInCloud(dirPath string, time time.Time) {
-	// first gather a list of ancestors
-	var ancestorCacheItems []*attrCacheItem
+
 	ancestorPath := internal.TruncateDirName(dirPath)
 	for ancestorPath != "" {
-		ancestorCacheItem, found := ac.cacheMap[ancestorPath]
-		if !(found && ancestorCacheItem.valid() && ancestorCacheItem.exists()) {
+
+		ancestorCacheItem, err := ac.cacheMap.get(ancestorPath)
+		if err != nil { //TODO: do more specific error check for attrCacheItem not existing
 			ancestorObjAttr := internal.CreateObjAttrDir(ancestorPath)
-			ancestorCacheItem = newAttrCacheItem(ancestorObjAttr, true, time)
-			ac.cacheMap[ancestorPath] = ancestorCacheItem
+			ac.cacheMap.insert(ancestorObjAttr, true, time)
 		}
-		ancestorCacheItems = append(ancestorCacheItems, ancestorCacheItem)
-		// speculatively set all ancestors as not in cloud storage
-		// all will be set correctly in the loop below
-		ancestorCacheItem.markInCloud(false)
-		// move on to the next ancestor
-		ancestorPath = getParentDir(ancestorPath)
-	}
-	// if we're at the root, no need to search (the root is always in the cloud)
-	if len(ancestorCacheItems) == 0 {
-		return
-	}
-	// search the cache to check whether each ancestor is in cloud storage
-cacheSearch:
-	for key, value := range ac.cacheMap {
-		// ignore items that are deleted, invalid, or directories that are not in cloud storage
-		if !value.exists() || !value.valid() || !value.isInCloud() {
-			continue
-		}
-		// iterate over ancestors, from the deepest up
-		prefixMatchFound := false
-	matchAncestors:
-		for ancestorIndex, ancestor := range ancestorCacheItems {
-			// don't visit ancestors that have already been updated
-			if ancestor.isInCloud() {
-				// if all ancestors have been updated, the entire search is done
-				if ancestorIndex == 0 {
-					break cacheSearch
+
+		var anyChildrenInCloud bool
+
+		for _, item := range ancestorCacheItem.children {
+			if item.exists() && item.isInCloud() {
+				anyChildrenInCloud = true
+				if anyChildrenInCloud {
+					break
 				}
-				break matchAncestors
-			}
-			// we already found that one ancestor is in the cloud
-			// so its ancestors are too
-			if prefixMatchFound {
-				ancestor.markInCloud(true)
-				continue matchAncestors
-			}
-			// check for a prefix match
-			prefix := dirToPrefix(ancestor.attr.Path)
-			if strings.HasPrefix(key, prefix) {
-				prefixMatchFound = true
-				// update this ancestor
-				ancestor.markInCloud(true)
 			}
 		}
+
+		if anyChildrenInCloud && !ancestorCacheItem.isInCloud() {
+			ancestorCacheItem.markInCloud(anyChildrenInCloud)
+		} else if !anyChildrenInCloud && ancestorCacheItem.isInCloud() {
+			ancestorCacheItem.markInCloud(anyChildrenInCloud)
```

Comment:
```suggestion
		if ancestorCacheItem.isInCloud() != anyChildrenInCloud {
			ancestorCacheItem.markInCloud(anyChildrenInCloud)
```
Since these do the same thing, it would be more compact to just ask if they're the same or not, and only have one `if`.